Expand `$ref` when assigning the `definition` field to `JsonSchemaValueException`

abravalheri opened the pull request:
This change should fix #141.

By expanding any `$ref` field in the schema definition that is available in the `JsonSchemaValueException` objects, we make it easier for the API consumers to provide more meaningful errors.

horejsek commented on fastjsonschema/generator.py:
```diff
@@ -245,8 +245,27 @@ def exc(self, msg, *args, rule=None):
+                return schema
+        if "properties" in definition:
+            return {
+                **definition,
```
Shouldn't definition here also recursively expand refs?

horejsek replied:
Tried to look around and found scenario here: https://json-schema.org/understanding-json-schema/reference/object.html#unevaluated-properties

```
{
  "type": "object",
  "properties": {
    "street_address": { "type": "string" },
    "city": { "type": "string" },
    "state": { "type": "string" },
    "type": { "enum": ["residential", "business"] }
  },
  "required": ["street_address", "city", "state", "type"],

  "if": {
    "type": "object",
    "properties": {
      "type": { "const": "business" }
    },
    "required": ["type"]
  },
  "then": {
    "properties": {
      "department": { "type": "string" }
    }
  },

  "unevaluatedProperties": false
}
```

In other words, if there are properties, there could be also if/then which can also contain refs. So I think that deserves tests on its own and do expansion here as well. Do you agree or maybe I have it wrong and this is enough? Maybe in either case test like that would be helpful so we can see its not needed as well. Do you think you can do that?

abravalheri replied:
Yeah, I agree that it is a good idea to add a test case for that.

Let me add it ;)

horejsek replied:
Thank you, appreciated!

abravalheri commented:
Thank you very much for the kind review @horejsek.

I don't know about [this question](https://github.com/horejsek/python-fastjsonschema/pull/142#discussion_r814725816), but I changed the code anyway as per your suggestion... I just had to add an additional condition to avoid failures, but then the tests pass.

abravalheri commented:
Question you be interested in another PR with improved error messages?

horejsek commented:
Sure, error messages need some improvements. There are also some open issues for it, so might be useful looking there as well. If you are willing to improve it, let me know what you have in mind if you want to discuss it first, and feel free to open PR for it.

abravalheri commented:
Thank you very much @horejsek, I added [tests](https://github.com/horejsek/python-fastjsonschema/pull/142/commits/64ac97e9d2d19154c4afc0836de9005a90140a23) using `$ref` inside `if-then-else`.


(I also added a regular [example](https://github.com/horejsek/python-fastjsonschema/pull/142/commits/cd13959062fd05ec79433baa4fcd75fd8afc0599) with `if-then-else`, just as a sanity check before working with `$ref` because I was having some problems with the syntax)

abravalheri commented:
I will have a look on the open issues and try to do some error message improvements in the future. Thank you very much!
